### DISCUSSION TITLE
TPM 2.0 with ECDSA key issue

### DISCUSSION THREAD

**Discussion by SKIPLALESSIVE** [strongswan/strongswan#2556] [Q&A]
Hello, 
I'm trying to setup a client server strongswan connection working fine with file but I'm trying to add a better security level with the TPM. 

I have an issue with the TPM 2.0 handle of my ECDSA key and I'm a little bit lost with it. 
I got a .pem ec private key registered in my tpm, working fine. 

But, when I try to use it with swanctl.conf I have a strange error at the IKE connection : 

no common hash algorithm found to create signature with ECDSA key

proposals = aes256gcm16-sha256-modp2048-ecp256

this is my proposals, when I tried to specifie ecdsa like this : 

proposals = aes256gcm16-sha256-modp2048-ecdsa

it return me the error that the config file is not correct. 
I also tried to use bad proposals for test purpose and I can see in the logs that the key is succesfully loaded. 

I cannot really see where is the error ? In the TPM ? the remote vpn ? my config ? 

Thx for helping me, 
Regards,

**Comment by strongX509** [comment]
What kind of ECC key did you create and store in the TPM 2.0?
* An ECDSA 256 bit key with the **`tpm2_createak`** parameters : **`-G ecc256 -g sha256 -s ecdsa `** or **`-G ecc -g sha256 -s ecdsa`** 
* An ECDSA 384 bit key with the **`tpm2_createak`** parameters:  **`-G ecc384 -g sha384 -s ecdsa `**

For the first case **`sha256`** would be ok in the **`proposals`** list, the second case would require **`sha384`**.

**Reply by SKIPLALESSIVE** [reply]
I try to add my own ECDSA256 key generate by my pki authority 

I type this commands after : 
tpm2_createprimary -Q -G ecc -g sha256 -C o -c parent.ctx
tpm2_import -G ecc -g sha256 -i test-XXXXXX.pem -C parent.ctx -u import_ecc_key.pub -r import_ecc_key.priv
tpm2_load -C parent.ctx -u import_ecc_key.pub -r import_ecc_key.priv -c key.ctx
tpm2_evictcontrol -C o -c key.ctx 0x81000002
And after I used 0x81000002 in the token_ak_ecc
Shoud I do something different ?

**Reply by strongX509** [reply]
I'm usually creating my ECC keys according to my TPM 2.0 tutorial:

1) Create the ECDSA Endorsement Key if it doesn't exist yet:
    https://docs.strongswan.org/docs/latest/tpm/tpm2.html#_ecc_endorsement_key

2) Create an ECDSA Attestation (i.e. a signature key) and persist it in the TPM 2.0:
    https://docs.strongswan.org/docs/latest/tpm/tpm2.html#_ecc_attestation_key

3) Generate a PKCS#10 Certificate Request using the strongSwan **`pki`** tool:
   https://docs.strongswan.org/docs/latest/tpm/tpm2.html#_ecc_certificate_request

4) Issue a host certificate based on the PKCS#10 Certificaton Request using your internal CA:
    https://docs.strongswan.org/docs/latest/tpm/tpm2.html#_ecc_attestation_key_certificate

5) Optionally persist the host certificate in the NV RAM of the TPM 2.0:
    https://docs.strongswan.org/docs/latest/tpm/tpm2.html#_storing_certificates_in_the_nv_ram

6) Use the ECDSA key and the matching host certificatge in a strongSwan IPsec connection
    https://docs.strongswan.org/docs/latest/tpm/tpm2Ike.html

Hope this helps

**Reply by SKIPLALESSIVE** [reply]
ok thx I will try this

**Reply by SKIPLALESSIVE** [reply]
Hi ! 
I tried this and I get few errors I cannot deal with : 
(I keep the name of your tutorial for test purpose ).

loading 'token_ak_ecc' failed: loading private key from token failed
loading connection 'test' failed: unable to load certificate: certs-tpm

 ---------------
I followed the tutorial you send to me, and that's return me this type of error. 

tpm2_createprimary -Q -G ecc -g sha256 -C o -c parent.ctx

tpm2_createek -G ecc -c 0x81010002 -u ek_ecc.pub

pki --print --type priv --keyid 0x81010002    // print the key correctly 

tpm2_createak -C 0x81010002 -G ecc -g sha256 -s ecdsa -c ak_ecc.ctx -u ak_ecc.pub -n ak_ecc.name

tpm2_evictcontrol -C o -c ak_ecc.ctx 0x81010004

pki --print --type priv --keyid 0x81010004 // print the key correctly 

sudo pki --req --type priv --keyid 0x81010004 --dn "XXXXXXXXX" --outform pem > ak_ecc_req.pem

tpm2_nvdefine 0x01800005 -C o -s 1001 -a 0x2000A

tpm2_nvwrite 0x01800005 -C o -i cert.der

in my swanctl/swanctl.conf : 


XXXXXXX
local {
                        auth = ecdsa // I also tried with pubkey without success 
                        certs-tpm {
                                        handle = 0x01800005
                                }
                        id = "XXXXXXX" // identical 
                                }


XXXXXX

secrets {
        token_ak_ecc {
        pin=""
        handle = 0x81010004
        }
}



Thx for your help

**Comment by strongX509** [comment]
In order to debug the problem I need a strongSwan log file. With strongswan as a systemd service. You get the last 200 lines of the log with
```
sudo journalctl -u strongswan -n 200
```

**Reply by SKIPLALESSIVE** [reply]
déc. 04 17:00:43 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Failed with result 'exit-code'.
déc. 04 17:00:43 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Failed to start strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl.
déc. 04 17:09:32 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Starting strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl...
déc. 04 17:09:32 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Main process exited, code=exited, status=66/NOINPUT
déc. 04 17:09:32 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Failed with result 'exit-code'.
déc. 04 17:09:32 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Failed to start strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl.
déc. 05 09:05:03 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Starting strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl...
déc. 05 09:05:03 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Main process exited, code=exited, status=66/NOINPUT
déc. 05 09:05:03 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Failed with result 'exit-code'.
déc. 05 09:05:03 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Failed to start strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl.
déc. 05 09:05:19 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Starting strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl...
déc. 05 09:05:19 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Main process exited, code=exited, status=66/NOINPUT
déc. 05 09:05:19 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Failed with result 'exit-code'.
déc. 05 09:05:19 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Failed to start strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl.
déc. 05 09:21:44 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Starting strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl...
déc. 05 09:21:44 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Main process exited, code=exited, status=66/NOINPUT
déc. 05 09:21:44 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Failed with result 'exit-code'.
déc. 05 09:21:44 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Failed to start strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl.
déc. 05 09:23:01 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Starting strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl...
déc. 05 09:23:01 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Main process exited, code=exited, status=66/NOINPUT
déc. 05 09:23:01 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Failed with result 'exit-code'.
déc. 05 09:23:01 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Failed to start strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl.
déc. 05 09:24:49 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Starting strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl...
déc. 05 09:24:49 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Main process exited, code=exited, status=66/NOINPUT
déc. 05 09:24:49 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: strongswan.service: Failed with result 'exit-code'.
déc. 05 09:24:49 mj-Standard-PC-i440FX-PIIX-1996 systemd[1]: Failed to start strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl.
lines 1-60


Uhm I think I got others errors.....
I compiled with hands and install packages so it's strange
also the strongswan-starter is running perfectly and the swanctl --load-all send answer me so ?

**Reply by strongX509** [reply]
The strongswan service might not start due syntax errors in **`swanctl.conf`**.  Have you still the set **`auth = ecdsa`** which is not a valid setting. Omit it or set the default value **`auth = pubkey`**. Also **`pin = ""`** is not acceptable. Just omit this parameter.

**Reply by SKIPLALESSIVE** [reply]
I did an entirely new VM with a fresh install and the tpm is correctly stored 
I don't really understand where I did an error but now both are stored ... 



Now I have an IKE error, that seems the client send the IKE AUTH and then TTL. 
 That's the log output : 
 
 
déc. 06 09:17:38 mj-Standard-PC-i440FX-PIIX-1996 charon-systemd[16272]: remote host is behind NAT
déc. 06 09:17:38 mj-Standard-PC-i440FX-PIIX-1996 charon-systemd[16272]: received cert request for "XXXXXPKI"
déc. 06 09:17:38 mj-Standard-PC-i440FX-PIIX-1996 charon-systemd[16272]: sending cert request for "XXXXPKI"
déc. 06 09:17:38 mj-Standard-PC-i440FX-PIIX-1996 charon-systemd[16272]: authentication of 'XXXXXCLIENCERT' (myself) with ECDSA_WITH_SHA256_DER successful
déc. 06 09:17:38 mj-Standard-PC-i440FX-PIIX-1996 charon-systemd[16272]: sending end entity cert "XXXXXCLIENCERT"
déc. 06 09:17:38 mj-Standard-PC-i440FX-PIIX-1996 charon-systemd[16272]: establishing CHILD_SA home{1}

generating IKE_AUTH request 1 [ IDi CERT N(INIT_CONTACT) CERTREQ IDr AUTH CPRQ(ADDR DNS) SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
déc. 06 09:17:38 mj-Standard-PC-i440FX-PIIX-1996 charon-systemd[16272]: sending packet: from 10.1.0.210[37809] to XXX.XXX.XXX.XXX[4500] (1596 bytes)

I randomise the port because the client and the server are behind NAT. 

I don't really understand with the ike failed because i don't receve failed auth message



`
/etc/swanct/swanctl.conf : 


 connections {
	home {
		version = 2
		remote_addrs = remote public ip 
		fragmentation = no
		mobike = yes
		vips = tunnel ip 
		local {
			auth = pubkey
			certs-tpm {
			handle = 0x01800004
			}
			id = "MY CERT "
	}
			
	remote {
	auth = pubkey 
	id = "PEER CERT"
	#id = %any
		}
		children {
			home {
			esp_proposals = default
			local_ts = tunnel ip declared on my ip a /30 
			remote_ts = 0.0.0.0/0
			start_action = start
			}
		}
	}

}
secrets {
token_ak_ecc {
	handle = 0x81010004 
}
}


/etc/strongswan.conf : 


charon {
	
	force_udp_encapsulation=yes 
	nat_keepalive=30
	port_nat_t=0
	ignore_rfc3947=no
	port=0
	install_virtual_ip=yes
	install_virtual_ip_on=ens18




	load_modular = yes
	plugins {
		include strongswan.d/charon/*.conf
	}
}

`

Thx a lot for your time and your help :)))

**Reply by strongX509** [reply]
The last message sent by the host is the  **`IKE_AUTH`** request:
```
generating IKE_AUTH request 1 [ IDi CERT N(INIT_CONTACT) CERTREQ IDr AUTH CPRQ(ADDR DNS) SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
sending packet: from 10.1.0.210[37809] to XXX.XXX.XXX.XXX[4500] (1596 bytes)
```
This IP packet has a size larger than 1596 bytes which is way higher then the MTU, so that the IP packet gets fragmented and is then probably discarded by a firewall in between, so that might be the reason that the **`IKE_AUTH`** response from the peer is never received.

Please set **`fragmentation = yes`**  in **`swanctl.conf`** so that the standard IKEv2 message fragmentation is applied.

**Reply by SKIPLALESSIVE** [reply]
I changed in both swanctl.conf, strongswan.conf and strongswan.d/charon.conf conf for fragmentation=yes and he keep sending 1596 bytes packet.

**Reply by strongX509** [reply]
Have you enabled fragmentation on the peer side, too? Only if both sides actively support fragmentation the IKE_AUTH messages get fragmented.  This is communicated via  **`FRAG_SUP`** notifications both in the **`IKE_SA_INIT`** request and subsequent **`IKE_SA_INIT`** response as shown in the following example:
```
* generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(REDIR_SUP) ]
  sending packet: from 192.168.0.1[500] to 192.168.0.2[500] (908 bytes)
* received packet: from 192.168.0.2[500] to 192.168.0.1[500] (281 bytes)
  parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(CHDLESS_SUP) ]
```

**Reply by SKIPLALESSIVE** [reply]
Hi ! 
Thx for your help, 
After further investigations I solved my issue 
Now I got another problem that I can struggle ton found something on google about : 

My child SA should be static, as I defined in swanctl.conf 
But the child sa failed and in my server logs I got Unable to assign IP : ModeConfig pool depleted 

here is my swanctl.conf file : 

connections {
	home {
	
		encap = yes
		version = 2
		remote_addrs = public ip 
		
		mobike = yes
		vips = ip i want in the tunnel and ip I declared on my machine 
		send_cert = always
		local {
			auth = pubkey
			certs = "XXX"
			id = "XXX"
	}
			
	remote {
	auth = pubkey 
	id = "XXX"

		}
		children {
			home {
			esp_proposals = default
			local_ts = ip i want in the tunnel and ip I declared on my machine 
			remote_ts = 0.0.0.0/0
			start_action = start
			}
		}
	}

}


and strongswan. conf : 



charon {

        force_udp_encapsulation=yes
        #fragmentation=yes 
        nat_keepalive=30
        port_nat_t=0
        ignore_rfc3947=no
        port=0
        install_virtual_ip=yes
        install_virtual_ip_on=ens18
        install_routes = no     

        load_modular = yes
        plugins {
                include strongswan.d/charon/*.conf
        }
}

include strongswan.d/*.conf

The client should get the ip i set on vips / local_ts right ?

**Reply by SKIPLALESSIVE** [reply]
: scheduling rekeying in 14076s
déc. 16 15:05:55 mj-Standard-PC-i440FX-PIIX-1996 charon-systemd[3857]: reauthentication already scheduled in 19851s
déc. 16 15:05:55 mj-Standard-PC-i440FX-PIIX-1996 charon-systemd[3857]: maximum IKE_SA lifetime 15516s
déc. 16 15:05:55 mj-Standard-PC-i440FX-PIIX-1996 charon-systemd[3857]: received INTERNAL_ADDRESS_FAILURE notify, no CHILD_SA built
déc. 16 15:05:55 mj-Standard-PC-i440FX-PIIX-1996 charon-systemd[3857]: failed to establish CHILD_SA, keeping IKE_SA



a sample of the error on the client
